lib/ansible_tower_client/api.rb

```diff
@@ -50,5 +50,53 @@ def method_missing(method_name, *args, &block)
     def respond_to_missing?(method, _include_private = false)
       instance.respond_to?(method)
     end
+
+    # Object class accessors patched for the appropriate version of the API
+
+    def ad_hoc_command_class
+      @ad_hoc_command_class ||= build_class("AdHocCommand")
+    end
+
+    def group_class
+      @group_class ||= build_class("Group")
+    end
+
+    def host_class
+      @host_class ||= build_class("Host")
+    end
+
+    def inventory_class
+      @inventory_class ||= build_class("Inventory")
+    end
+
+    def inventory_source_class
+      @inventory_source_class ||= build_class("InventorySource")
+    end
+
+    def inventory_update_class
+      @inventory_update_class ||= build_class("InventoryUpdate")
+    end
+
+    def job_class
+      @job_class ||= build_class("Job")
+    end
+
+    def job_template_class
+      @job_template_class ||= build_class("JobTemplate").tap do |klass|
+        if Gem::Version.new(version).between?(Gem::Version.new(2), Gem::Version.new(3))
+          require "ansible_tower_client/v2_patches/job_template_launch"
+          klass.prepend(AnsibleTowerClient::V2Patches::JobTemplateLaunch)
+        end
+      end
+    end
+
+    private
+
+    def build_class(model_name)
+      base_model_constant = AnsibleTowerClient.const_get(model_name)
+      Class.new(base_model_constant).tap do |klass|
```

Comment:
I'm concerned this could get expensive in a similar way to how creating OpenStructs en masse used to be expensive.  I'm thinking it makes more sense to create all of these class types up front using normal inheritance, and then just picking the right one at run time, instead of dynamically creating them.  I'm pretty sure that would be possible, and not overly complex.